shell/browser/api/atom_api_data_pipe_holder.cc

```diff
@@ -0,0 +1,185 @@
+
+namespace electron {
+
+namespace api {
+
+namespace {
+
+// Incremental ID.
+int g_next_id = 0;
+
+// Map that manages all the DataPipeHolder objects.
+KeyWeakMap<std::string> g_weak_map;
+
+// Utility class to read from data pipe.
+class DataPipeReader {
+ public:
+  DataPipeReader(util::Promise<v8::Local<v8::Value>> promise,
+                 network::mojom::DataPipeGetterPtr data_pipe_getter)
+      : promise_(std::move(promise)),
+        data_pipe_getter_(std::move(data_pipe_getter)),
+        handle_watcher_(FROM_HERE,
+                        mojo::SimpleWatcher::ArmingPolicy::MANUAL,
+                        base::SequencedTaskRunnerHandle::Get()) {
+    // Get a new data pipe and start.
+    mojo::DataPipe data_pipe;
+    data_pipe_getter_->Read(std::move(data_pipe.producer_handle),
+                            base::BindOnce(&DataPipeReader::ReadCallback,
+                                           weak_factory_.GetWeakPtr()));
+    data_pipe_ = std::move(data_pipe.consumer_handle);
+    handle_watcher_.Watch(data_pipe_.get(), MOJO_HANDLE_SIGNAL_READABLE,
+                          base::BindRepeating(&DataPipeReader::OnHandleReadable,
+                                              weak_factory_.GetWeakPtr()));
+  }
+
+  ~DataPipeReader() = default;
+
+ private:
+  // Callback invoked by DataPipeGetter::Read.
+  void ReadCallback(int32_t status, uint64_t size) {
+    if (status != net::OK) {
+      OnFailure();
+      return;
+    }
+    buffer_.resize(size);
+    head_ = &buffer_.front();
+    remaining_size_ = size;
+    handle_watcher_.ArmOrNotify();
+  }
+
+  // Called by |handle_watcher_| when data is available or the pipe was closed,
+  // and there's a pending Read() call.
+  void OnHandleReadable(MojoResult result) {
+    if (result != MOJO_RESULT_OK) {
+      OnFailure();
+      return;
+    }
+
+    // Read.
+    uint32_t length = remaining_size_;
+    result = data_pipe_->ReadData(head_, &length, MOJO_READ_DATA_FLAG_NONE);
+    if (result == MOJO_RESULT_OK) {  // success
+      remaining_size_ -= length;
+      head_ += length;
+      if (remaining_size_ == 0)
+        OnSuccess();
+    } else if (result == MOJO_RESULT_SHOULD_WAIT) {  // IO pending
+      handle_watcher_.ArmOrNotify();
+    } else {  // error
+      OnFailure();
+    }
+  }
+
+  void OnFailure() {
+    promise_.RejectWithErrorMessage("Could not get blob data");
+    delete this;
+  }
+
+  void OnSuccess() {
+    // Pass the buffer to JS.
+    //
+    // Note that the lifetime of the native buffer belongs to us, and we will
+    // free memory when JS buffer gets garbage collected.
+    v8::Locker locker(promise_.isolate());
+    v8::HandleScope handle_scope(promise_.isolate());
+    v8::Local<v8::Value> buffer =
+        node::Buffer::New(promise_.isolate(), &buffer_.front(), buffer_.size(),
+                          &DataPipeReader::FreeBuffer, this)
+            .ToLocalChecked();
+    promise_.Resolve(buffer);
+
```

Comment:
Shouldn't the `DataPipeReader` class also be deleted here ?

Reply:
The `Buffer` passed to JS reuses the C++ buffer instead of copying the data, so `DataPipeReader` is kept alive until the JS buffer gets garbage collected.

Reply:
Ahh I didn't check that carefully, looks good then. Thanks!